otcextensions/sdk/dms/v1/message.py

```diff
@@ -9,23 +9,112 @@
 # WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 # License for the specific language governing permissions and limitations
 # under the License.
+from openstack import exceptions
 from openstack import resource
-from openstack import _log
 
-from otcextensions.sdk.dms.v1 import _base
 
-_logger = _log.setup_logging('openstack')
+class Message(resource.Resource):
 
+    resource_key = 'message'
+    base_path = '/queues/%(queue_id)s/groups/%(group_id)s/messages'
 
-class Message(_base.Resource):
+    allow_list = True
+
+    _query_mapping = resource.QueryParameters(
+        'max_msgs', 'time_wait', 'ack_wait'
+    )
+
+    #: Queue id
+    queue_id = resource.URI('queue_id')
+    #: Group ID (part of URL for listing)
+    group_id = resource.URI('group_id')
+
+    # Properties
+    #: Attributes
+    attributes = resource.Body('attributes', type=dict)
+    #: Message body
+    body = resource.Body('body')
+    #: Error info
+    error = resource.Body('error')
+    #: Error code
+    error_code = resource.Body('error_code')
+    #: Message handler (ID)
+    handler = resource.Body('handler', alternate_id=True)
+    #: State (0 - success)
+    state = resource.Body('state', type=int)
+
+    def _collect_attrs(self, attrs):
+        """ Save remaining attributes under "attributes" attribute
+        """
+        body = self._consume_body_attrs(attrs)
+        header = self._consume_body_attrs(attrs)
+        uri = self._consume_body_attrs(attrs)
+        body['attributes'] = attrs
+        computed = self._consume_attrs(self._computed_mapping(), attrs)
+
+        return body, header, uri, computed
+
+    @classmethod
+    def list(cls, session, paginated=True, base_path=None,
+             allow_unknown_params=False, **params):
+
+        microversion = cls._get_microversion_for_list(session)
+
+        if base_path is None:
+            base_path = cls.base_path
+        params = cls._query_mapping._validate(
+            params, base_path=base_path,
+            allow_unknown_params=allow_unknown_params)
+        query_params = cls._query_mapping._transpose(params, cls)
+        uri = base_path % params
+
+        params = cls._query_mapping._validate(
+            params, base_path=base_path,
+            allow_unknown_params=allow_unknown_params)
+        query_params = cls._query_mapping._transpose(params, cls)
+        uri = base_path % params
+
+        while uri:
+            # Copy query_params due to weird mock unittest interactions
+            response = session.get(
+                uri,
+                headers={"Accept": "application/json"},
+                params=query_params.copy(),
+                microversion=microversion)
+            exceptions.raise_from_response(response)
+            data = response.json()
```

Comment:
do we need data? cannot we use directly resources?

Reply:
again - copy from upstream. We can do lot's of optimizations, but they make it harder to backport changes from upstream